plutarch-test/src/Plutarch/Test.hs

```diff
@@ -0,0 +1,139 @@
+  p1 <- printTermEvaluated x
+  p2 <- printTermEvaluated y
+  p1 `shouldBe` p2
+
+{- Like `@?=` but for Plutarch terms -}
+(#@?=) :: forall (a :: PType) (b :: PType). ClosedTerm a -> ClosedTerm b -> Assertion
+(#@?=) = pshouldBe
+
+eval :: Scripts.Script -> IO Scripts.Script
+eval s = case evaluateScript s of
+  Left e -> assertFailure $ "Script evaluation failed: " <> show e
+  Right (_, _, x') -> pure x'
+
+{- Like `printTerm` but the prints evaluated output of it -}
+printTermEvaluated :: ClosedTerm a -> IO String
+printTermEvaluated = fmap printScript . eval . compile
+
+{- Asserts the term to be true -}
+passert :: forall (a :: PType). ClosedTerm a -> Assertion
+passert p = p #@?= pcon PTrue
+
+psucceeds :: forall (a :: PType). ClosedTerm a -> Assertion
+psucceeds p =
+  case evaluateScript (compile p) of
+    Left _ -> assertFailure $ "Term failed to evaluate"
+    Right _ -> pure ()
+
+{- Asserts the term evaluates without success -}
+pfails :: forall (a :: PType). ClosedTerm a -> Assertion
+pfails p = do
+  case evaluateScript (compile p) of
+    Left _ -> pure ()
+    Right _ -> assertFailure $ "Term succeeded"
+
+{- Whether to run all or a particular golden test -}
+data PlutarchGolden
+  = All
+  | Bench
+  | PrintTerm
+  deriving stock (Eq, Show)
+
+hasBenchGolden :: PlutarchGolden -> Bool
+hasBenchGolden = \case
+  PrintTerm -> False
+  _ -> True
+
+hasPrintTermGolden :: PlutarchGolden -> Bool
+hasPrintTermGolden = \case
+  Bench -> False
+  _ -> True
+
+{- Run golden tests on the given Plutarch program -}
+golden :: PlutarchGolden -> ClosedTerm a -> Spec
+golden pg = golden' pg Nothing
+
+-- | Make golden tests for the given Plutarch program.
+golden' :: forall a. PlutarchGolden -> Maybe String -> ClosedTerm a -> Spec
+golden' pg mk p =
+  goldens' pg mk [("0", popaque p)]
+
+{- | Like `golden` but for multiple programs
+
+  Multiple programs use a single golden file. Each output separated from the
+  keyword with a space.
+-}
+goldens :: PlutarchGolden -> [(String, ClosedTerm a)] -> Spec
+goldens pg = goldens' pg Nothing
+
+goldens' :: PlutarchGolden -> Maybe String -> [(String, ClosedTerm a)] -> Spec
+goldens' pg mk ps = do
+  testAncestors <- fmap (drop 1 . reverse) $ getTestDescriptionPath
+  let name = T.unpack $ T.intercalate "." testAncestors
+      goldenKey = maybe "golden" (<> ".golden") mk
+  describe goldenKey $ do
+    let k = maybe "" ("." <>) mk
+        nUplc = name <> k <> ".uplc.golden"
+        nBench = name <> k <> ".bench.golden"
+    -- Golden test for UPLC
+    when (hasPrintTermGolden pg) $
+      it "uplc" $
+        pureGoldenByteStringFile ("goldens" </> nUplc) $
+          multiGolden ps $ \p ->
+            T.pack $ printTerm p
+    -- Golden test for Plutus benchmarks
+    when (hasBenchGolden pg) $
+      it "bench" $
+        pureGoldenByteStringFile ("goldens" </> nBench) $
+          multiGolden ps $ \p ->
+            TL.toStrict $ Aeson.encodeToLazyText $ benchmarkScript' $ compile p
```

Comment:
I would use `sydtest-aeson`'s function for this.

Reply:
Some goldens contain multiple entries, eg:

```
❯ cat plutarch-test/goldens/bool.pnot.bench.golden 
lam {"exBudgetCPU":29873,"exBudgetMemory":200,"scriptSizeBytes":14}
app {"exBudgetCPU":387150,"exBudgetMemory":1401,"scriptSizeBytes":16}
```

Looks like `sydtest-aeson` shows a JSON-friendly diff. Can we get that for the above case? `pureGoldenJSONValueFileMany :: (Show a, Eq a, FromJSON a, ToJSON a) => FilePath -> Map k a -> GoldenTest a`?

Reply:
The reason why some goldens have multiple JSON values in them is to group benchmarks for closely related programs in the same file, and thus avoid having gazillion golden files most of which corresponding to tiny programs.

Reply:
@srid if we have the name for each of those things in the multi-golden, we could make a json object out of them, otherwise a list...

Reply:
Using json object of assoc list (`[String, a]`), `sydtest-aeson` is producing:

<img width="459" alt="image" src="https://user-images.githubusercontent.com/3998/153913492-6a456353-8e4c-4c1a-ac15-fa38496ad199.png">

compared to the current:

<img width="1180" alt="image" src="https://user-images.githubusercontent.com/3998/153913818-95e557a9-8d98-4b5c-be19-6444c2bd922a.png">

I don't like how the latter (ie., current behaviour) puts both the entries (`app` and `lam`) in the _same_ line; at the same time I find the former to be too verbose. 🤔

Reply:
I like the `pureGoldenTextFile` one best. But it's going to be messed up if we change the keys to a different name. Up to you.

Reply:
Keys and order shouldn't change unless developers purposefully change their name/order in the test source. The above golden is for:

```haskell
      goldens
        All
        [ ("lam", popaque pnot)
        , ("app", popaque $ pnot #$ pcon PTrue)
        ]
```